qiskit/transpiler/passes/basis/basis_translator.py

```diff
@@ -312,100 +398,75 @@ def _basis_search(equiv_lib, source_basis, target_basis, heuristic):
-            if gate_name in target_basis:
-                continue
-
-            equivs = equiv_lib._get_equivalences((gate_name, gate_num_qubits))
-
-            basis_remain = current_basis - {(gate_name, gate_num_qubits)}
-            neighbors = [
+    source_basis = {
+        (gate_name, gate_num_qubits)
+        for gate_name, gate_num_qubits in source_basis
+        if gate_name not in target_basis
+    }
+
+    # if source basis is empty, no work to be done.
+    if not source_basis:
+        return []
+
+    all_gates_in_lib = set()
+
+    graph = retworkx.PyDiGraph()
+    nodes_to_indices = dict()
+    num_gates_for_rule = dict()
+
+    def lazy_setdefault(key):
+        if key not in nodes_to_indices:
+            nodes_to_indices[key] = graph.add_node(key)
+        return nodes_to_indices[key]
+
+    rcounter = 0  # running sum of the number of equivalence rules in the library.
+    for key in equiv_lib._get_all_keys():
+        target = lazy_setdefault(key)
+        all_gates_in_lib.add(key)
+        for equiv in equiv_lib._get_equivalences(key):
+            sources = {
+                Key(name=gate.name, num_qubits=len(qargs)) for gate, qargs, _ in equiv.circuit
+            }
+            all_gates_in_lib |= sources
+            edges = [
                 (
-                    frozenset(
-                        basis_remain
-                        | {(inst.name, inst.num_qubits) for inst, qargs, cargs in equiv.data}
-                    ),
-                    params,
-                    equiv,
+                    lazy_setdefault(source),
+                    target,
+                    {"index": rcounter, "rule": equiv, "source": source},
                 )
-                for params, equiv in equivs
+                for source in sources
             ]
 
-            # Weight total path length of transformation weakly.
-            tentative_cost_from_source = cost_from_source[current_basis] + 1e-3
-
-            for neighbor, params, equiv in neighbors:
-                if neighbor in closed_set:
-                    continue
-
-                if tentative_cost_from_source >= cost_from_source[neighbor]:
-                    continue
-
-                open_set.add(neighbor)
-                came_from[neighbor] = (current_basis, gate_name, gate_num_qubits, params, equiv)
-                cost_from_source[neighbor] = tentative_cost_from_source
-                est_total_cost[neighbor] = tentative_cost_from_source + heuristic(
-                    neighbor, target_basis
-                )
-                heappush(open_heap, (est_total_cost[neighbor], next(basis_count), neighbor))
-
-    return None
+            num_gates_for_rule[rcounter] = len(sources)
+            graph.add_edges_from(edges)
+            rcounter += 1
+
+    # This is only neccessary since gates in target basis are currently reported by
+    # their names and we need to have in addition the number of qubits they act on.
```

Comment:
If a target is set we can get this via:

```
gate_num_qubits = target.operation_from_name(name).num_qubits
```

but that doesn't cover the non-target path.